### PR TITLE
Ft/test elastic ilm

### PR DESCRIPTION
**Pull request by thepsalmist**
Docker compose test implementation of managing Elasticsearch Indexes through an ILM policy.

1. We introduce a conf/elaticsearch directory to add our configuration files
2. Indexing at the importer now done to an alias that maps to all indices created by the ILM & index template
2. The Elasticsearch setup is done via an `elasticsearch-conf` service (built from common worker dockerfile) 
 
By default ILM checks every 10 minutes if there’s any action to execute. In our example we had to use very small intervals (seconds, minutes), for testing.
Execute the following command to execute ILM

`
PUT _cluster/settings
{
  "transient": {
    "indices.lifecycle.poll_interval": "15s" 
  }
}
`

### CODE REVIEW

File: indexer/workers/importer.py [hunk 1]
```diff
@@ -167,11 +89,12 @@ def import_story(
         if data:
             url = str(data.get("url"))
             url_hash = hashlib.sha256(url.encode("utf-8")).hexdigest()
-            publication_date = str(data.get("publication_date"))
+            # publication_date = str(data.get("publication_date"))
             # Add the indexed_date with today's date in ISO 8601 format
             indexed_date = datetime.now().isoformat()
             data = {**data, "indexed_date": indexed_date}
-            target_index = self.index_routing(publication_date)
+            # target_index = self.index_routing(publication_date)
```
Comment (kilemensi) [inline indexer/workers/importer.py:96]:
```suggestion
```

File: indexer/workers/importer.py [hunk 2]
```diff
@@ -167,11 +89,12 @@ def import_story(
         if data:
             url = str(data.get("url"))
             url_hash = hashlib.sha256(url.encode("utf-8")).hexdigest()
-            publication_date = str(data.get("publication_date"))
+            # publication_date = str(data.get("publication_date"))
```
Comment (kilemensi) [inline indexer/workers/importer.py:92]:
```suggestion
```

File: conf/elasticsearch/create_ilm_policy.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (kilemensi) [inline conf/elasticsearch/create_ilm_policy.json:1]: Are these configs for PROD or STAGING/DEV?

File: indexer/scripts/elastic-conf.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+
+class ElasticConf(ElasticMixin, App):
+    def main_loop(self) -> None:
+        es = self.elasticsearch_client()
+        assert es.ping(), "Failed to connect to Elasticsearch"
+
+        index_template_path = "./conf/elasticsearch/create_index_template.json"
+        ilm_policy_path = "./conf/elasticsearch/create_ilm_policy.json"
+        initial_index_template = "./conf/elasticsearch/create_initial_index.json"
```
Comment (kilemensi) [inline indexer/scripts/elastic-conf.py:29]: Isn't it better to have conf env var that can be reused e.g. ELASTICSEARCH_CONF_DIR, then:
```suggestion
        index_template_path = f"{ELASTICSEARCH_CONF_DIR}/create_index_template.json"
        ilm_policy_path = f"{ELASTICSEARCH_CONF_DIR}/create_ilm_policy.json"
        initial_index_template = f"{ELASTICSEARCH_CONF_DIR}/create_initial_index.json"
```

Or something like that.

File: indexer/scripts/elastic-conf.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+        index_template_created = self.create_index_template(es, index_template_path)
+        ilm_policy_created = self.create_ilm_policy(es, ilm_policy_path)
+        alias_created = self.create_initial_index(es, initial_index_template)
+        # snapshot_policy_created = self.create_snapshot_policy(es, snapshot_policy_path)
+
+        if index_template_created and ilm_policy_created and alias_created:
+            logger.info("All ES configurations applied successfully.")
+        else:
+            logger.error("One or more configurations failed. Check logs for details.")
```
Comment (kilemensi) [inline indexer/scripts/elastic-conf.py:40]: Shouldn't the app exit here if there are configuration errors?

File: indexer/scripts/elastic-conf.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+    def create_index_template(self, es: Elasticsearch, file_path: str) -> bool:
+        with open(file_path, "r") as file:
+            data = file.read()
+
+        json_data = json.loads(data)
```
Comment (kilemensi) [inline indexer/scripts/elastic-conf.py:46]: These are repeated in every method... may be there should be just one private method to load JSON from file path?

File: indexer/scripts/elastic-conf.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        name = json_data.get("name")
+        template = json_data.get("template")
+        index_patterns = json_data.get("index_patterns")
+
+        # Check if the index template already exists
+        if es.indices.exists_index_template(name=name):
+            logger.info("Index template ':%s' already exists. Skipping creation.", name)
+            return True
```
Comment (kilemensi) [inline indexer/scripts/elastic-conf.py:54]: Doesn't this mean the template will never be updated even if we update the contents of the JSON file?

**Review by kilemensi — COMMENTED** [review]
👍🏽 

--

I'm assuming you were referring to this setup as production: https://github.com/CodeForAfrica/story-indexer/blob/d48623f1ab3c80ecef4682e24275ceacfc16a2f0/docker/deploy.sh#L222-L224

Can you may be clarify how this would work (besides the number of containers)? For example, what does 1 replica mean? Is it an additional node, or does each node act as a replica of another node in the cluster?

**Comment by kilemensi** [comment]
PS: May be I misheard you but I thought you said the Story fix was also in this PR @thepsalmist. I didn't see it.

**Comment by thepsalmist** [comment]
Sorry the updated version of this is here https://github.com/mediacloud/story-indexer/pull/211, will apply the suggestions/feedback on that PR

**Comment by kilemensi** [comment]
> Sorry the updated version of this is here [mediacloud#211](https://github.com/mediacloud/story-indexer/pull/211), will apply the suggestions/feedback on that PR

Cool, then lets move the Story fix into it's own PR so it can be review and merge quickly @thepsalmist

File: indexer/scripts/elastic-conf.py (hunk 7 above)
Comment (thepsalmist) [reply indexer/scripts/elastic-conf.py:54]: Cool `put_index_template` allows for create/update without raising conflict

**Comment by thepsalmist** [comment]
Updates resolved on https://github.com/mediacloud/story-indexer/pull/211

**Pull request merged**